### DISCUSSION TITLE
Local Search in CP-SAT

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#3387] [CP-SAT questions] [answered]
Hi Experts,

Do we have local search algorithms[NOT LNS] and parameter to control it in CP-SAT? Thanks.

John

**Comment by lperron** [comment, marked as answer]
No. For a simple reason, this is hard.

A good illustration is that there are no LS implementation in MIP solvers, and CP-SAT is a superset of a MIP solver.

We welcome any idea that has a decent chance of success.

**Reply by lperron** [reply]
We now have. It is violation based, and inspired by the paper [feasibility jump](https://link.springer.com/article/10.1007/s12532-023-00234-8). It is on be default starting v9.7 to find initial solution. You can specify how many workers to allocate with the `num_violation_ls` parameter. The default is 0. It is disabled for models containing no_overlap_2d. It is only optimized for the linear case (linear, bool_or, bool_and, at_most_one, exactly_one).

**Reply by ofek** [reply]
It looks like there is a new state-of-the-art technique that performs better than feasibility jump in all circumstances https://drops.dagstuhl.de/storage/00lipics/lipics-vol307-cp2024/LIPIcs.CP.2024.19/LIPIcs.CP.2024.19.pdf

**Comment by lperron** [comment]
I ran some comparison. it is not better than what we have in CP-SAT.

And incidentally, they did not compare to violationLS at the conference.

**Reply by ofek** [reply]
Oh very interesting, thanks for the heads up! I'm in the process of researching optimization software and trying to figure out what is most performant (use case is healthcare employee scheduling). All of the quotes I've gotten for proprietary solutions have been extraordinarily high (Gurobi, Cardinal Optimizer, Hexaly, Timefold) and I can't find any benchmarks for the proprietary solutions except for Hexaly which they publish themselves. I found [these benchmarks](https://plato.asu.edu/bench.html) from a researcher but they don't include proprietary solutions nor or-tools.

**Comment by lperron** [comment]
Look at shift_schedulinh_sat.py example.

Very popular in nurse rostering

**Reply by Mizux** [reply]
https://github.com/google/or-tools/blob/stable/examples/python/shift_scheduling_sat.py

**Reply by ofek** [reply]
Yes thank you. That was the first thing that I tried but on my machine it takes approximately 10 seconds for that very small example.
